### PR TITLE
Add mTLS and Trust/KeyManager Credentials API

### PR DESCRIPTION
**Pull request by ejona86**
It feels great cleaning up the test code.

### CODE REVIEW

File: api/src/main/java/io/grpc/TlsChannelCredentials.java [hunk 1]
```diff
@@ -34,9 +44,82 @@ public static ChannelCredentials create() {
   }
 
   private final boolean fakeFeature;
+  private final byte[] certificateChain;
+  private final byte[] privateKey;
+  private final String privateKeyPassword;
+  private final List<KeyManager> keyManagers;
+  private final byte[] rootCertificates;
```
Comment (sanjaypujare) [inline api/src/main/java/io/grpc/TlsChannelCredentials.java:51]: Just started the review so there might something later that would answer this question: why use `byte[]` instead of `PrivateKey`, `Certificate` etc? AFAICS there is no additional dependency or netty vs okhttp issue.

**Comment by jiangtaoli2016** [comment]
cc @ZhenLian FYI.

File: api/src/main/java/io/grpc/TlsChannelCredentials.java (hunk 1 above)
Comment (ejona86) [reply api/src/main/java/io/grpc/TlsChannelCredentials.java:51]: We're already using byte[] on server-side for key/cert. We could convert byte[] to a Certificate pretty easily. Although for Netty, we don't actually want to do that. For Netty, there's an advantage of using byte[] as then we can pass it to boringssl and have boringssl manage the data (which allows some faster code paths). Given the Java objects PrivateKey and Certificate, we could serialize them for Netty, but it is just more of a pain and we'd have to write that code.

PrivateKey however is different. There's no built-in parser for parsing a der-encoded key. It isn't _that_ bad, but requires a base64 parser and some levels of hackery. Netty has that implemented. But we don't have an implementation for OkHttp; we'd need to write it.

Users tend to have bytes and not Certificate or PrivateKey. So by aligning the API to what the user's have we make the Credentials more of a shim and move the more complex conversion logic into the transports.

File: api/src/main/java/io/grpc/TlsChannelCredentials.java [hunk 2]
```diff
@@ -119,6 +242,126 @@ public Builder requireFakeFeature() {
...
+      this.privateKeyPassword = privateKeyPassword;
+      return this;
+    }
+
+    /**
+     * Have the provided key manager select the client's identity. Although multiple are allowed,
+     * only the first instance implementing a particular interface is used. So generally there will
+     * just be a single entry and it implements {@link javax.net.ssl.X509KeyManager}.
+     */
+    public Builder keyManager(KeyManager... keyManagers) {
```
Comment (sanjaypujare) [inline api/src/main/java/io/grpc/TlsChannelCredentials.java:305]: This is quite similar to `SslContextBuilder` and the intent might be to provide similar functionality.

- how about a method to accept a `KeyManagerFactory` like the SslContextBuilder? I see that you are anyway wrapping this list in a `FixedKeyManagerFactory`
- then just use a `KeyManagerFactory` inside and use a `KeyManagerFactoryWrapper` or `FixedKeyManagerFactory` to support these other KeyManager setters.

Comment (ejona86) [reply api/src/main/java/io/grpc/TlsChannelCredentials.java:305]: I'm making the grpc-netty implementation more complicated to workaround a useless Netty-ism (that should really be fixed) while making the user-facing API easier to use.

Java [accepts the managers](https://docs.oracle.com/javase/8/docs/api/javax/net/ssl/SSLContext.html#init-javax.net.ssl.KeyManager:A-javax.net.ssl.TrustManager:A-java.security.SecureRandom-), not the factories. The factories are actually a giant pain, because they require a Provider and an algorithm and that stuff doesn't actually matter to any consumer of the object.

KeyManager and TrustManager are much easier for users to implement their own. For example, for the "advanced TLS" APIs, they will provide a custom KeyManager and TrustManager implementation. I don't want to _force_ the user to convert those to a Factory.

Netty uses the Factories in its API, and it seems clearly a mistake. The `Fixed*Factory`s are a workaround, and they do weird things. I should really be changing Netty to use Manager lists internally; I've checked and I only saw calls the `get*Managers()` method. That would remove the need for _anyone_ to use `*ManagerFactoryWrapper` that is in Netty or the `Fixed*ManagerFactory` that I add here.

File: api/src/main/java/io/grpc/TlsChannelCredentials.java [hunk 3]
```diff
@@ -119,6 +242,126 @@ public Builder requireFakeFeature() {
...
+    /**
+     * Use the provided certificate chain and possibly-encrypted private key as the client's
+     * identity. Generally they should be PEM-encoded and the key is a PKCS#8 key. If the private
+     * key is unencrypted, then password must be {@code null}.
+     */
+    public Builder keyManager(File certChain, File privateKey, String privateKeyPassword)
+        throws IOException {
+      InputStream certChainIs = new FileInputStream(certChain);
+      try {
+        InputStream privateKeyIs = new FileInputStream(privateKey);
```
Comment (sanjaypujare) [inline api/src/main/java/io/grpc/TlsChannelCredentials.java:263]: I believe we can use "try-with-resources" here and similar places?

Comment (ejona86) [reply api/src/main/java/io/grpc/TlsChannelCredentials.java:263]: I think try-with-resources is still off limits due to Android. (It may actually be safe now, but there's complications with suppressed exceptions and there's multiple Android toolchains and tracking each feature is hard. All this will change soon anyway once we drop Java 7 support.)

File: api/src/main/java/io/grpc/TlsServerCredentials.java [hunk 4]
```diff
@@ -196,36 +291,130 @@ public Builder keyManager(File certChain, File privateKey, String privateKeyPass
...
+
+    private void clearTrustManagers() {
+      this.rootCertificates = null;
+      this.trustManagers = null;
+    }
+
     /** Construct the credentials. */
     public ServerCredentials build() {
-      if (certificateChain == null) {
+      if (certificateChain == null && keyManagers == null) {
```
Comment (sanjaypujare) [inline api/src/main/java/io/grpc/TlsServerCredentials.java:396]: Should you also check that if ClientAuth is REQUIRE then the trustManager cannot be null?

File: android-interop-testing/src/main/java/io/grpc/android/integrationtest/TesterOkHttpChannelBuilder.java [hunk 5]
```diff
@@ -45,48 +37,26 @@ public static ManagedChannel build(
...
+          credentials = TlsChannelCredentials.newBuilder().trustManager(testCa).build();
+        } catch (Exception e) {
+          throw new RuntimeException(e);
+        }
       }
     } else {
       credentials = InsecureChannelCredentials.create();
     }
 
-    ManagedChannelBuilder<?> channelBuilder = Grpc.newChannelBuilderForAddress(
```
Comment (sanjaypujare) [inline android-interop-testing/src/main/java/io/grpc/android/integrationtest/TesterOkHttpChannelBuilder.java:57]: WIth this change there will be no interop test for `Grpc.newChannelBuilderForAddress` on Android with OkHttp?

File: interop-testing/src/test/java/io/grpc/testing/integration/Http2NettyTest.java [hunk 6]
```diff
@@ -68,16 +65,15 @@
...
   protected NettyChannelBuilder createChannelBuilder() {
     try {
+      ChannelCredentials channelCreds = TlsChannelCredentials.newBuilder()
+          .keyManager(TestUtils.loadCert("client.pem"), TestUtils.loadCert("client.key"))
+          .trustManager(TestUtils.loadCert("ca.pem"))
+          .build();
       NettyChannelBuilder builder = NettyChannelBuilder
-          .forAddress(TestUtils.testServerAddress((InetSocketAddress) getListenAddress()))
+          .forAddress("localhost", ((InetSocketAddress) getListenAddress()).getPort(), channelCreds)
+          .overrideAuthority(TestUtils.TEST_SERVER_HOST)
```
Comment (sanjaypujare) [inline interop-testing/src/test/java/io/grpc/testing/integration/Http2NettyTest.java:74]: Why is this needed now? The setup hasn't changed - only using the new creds API AFAICS

File: android-interop-testing/src/main/java/io/grpc/android/integrationtest/TesterOkHttpChannelBuilder.java (hunk 5 above)
Comment (ejona86) [reply android-interop-testing/src/main/java/io/grpc/android/integrationtest/TesterOkHttpChannelBuilder.java:57]: My earlier creds change swapped to `Grpc.newChannelBuilderForAddress`. I'm not too concerned about whether it is used/not. Before my earlier change [it was using `ManagedChannelBuilder.forAddress()`](https://github.com/grpc/grpc-java/blob/5a687e3da8fce41b51b9ffe73c8bba6f9c8a053e/android-interop-testing/src/main/java/io/grpc/android/integrationtest/TesterOkHttpChannelBuilder.java), and then casting the result to OkHttpChannelBuilder.

The only reason I'm using OkHttpChannelBuilder is to force okhttp to be used. I guess I could just do an instanceof check on the value returned by `Grpc.newChannelBuilderForAddress`... Sure. I'll do that.

File: interop-testing/src/test/java/io/grpc/testing/integration/Http2NettyTest.java (hunk 6 above)
Comment (ejona86) [reply interop-testing/src/test/java/io/grpc/testing/integration/Http2NettyTest.java:74]: Take a [look at `TestUtils.testServerAddress()`](https://github.com/grpc/grpc-java/blob/3752b9e36590793ca1695faacabfe561a1b68305/testing/src/main/java/io/grpc/internal/testing/TestUtils.java#L70). It creates the address so the IP address points to localhost, but then injects TEST_SERVER_HOST as the "hostname" of the `InetSocketAddress`. That hostname is used as the default authority.

We currently don't have a `forAddress(SocketAddress, ChannelCreds)` API (which would be fair to add), but there's no good reason to continue using this approach. This code probably predates NameResolvers and `overrideAuthority()`. We actually don't want to use the `forAddress(SocketAddress)` often, as it disables the NameResolver.

File: netty/src/main/java/io/grpc/netty/FixedTrustManagerFactory.java [hunk 7]
```diff
@@ -0,0 +1,53 @@
...
+import java.util.Collections;
+import java.util.List;
+import javax.net.ssl.ManagerFactoryParameters;
+import javax.net.ssl.TrustManager;
+import javax.net.ssl.TrustManagerFactory;
+import javax.net.ssl.TrustManagerFactorySpi;
+
+/** A TrustManagerFactory that returns a fixed list of key managers. */
+final class FixedTrustManagerFactory extends TrustManagerFactory {
+  public FixedTrustManagerFactory(List<TrustManager> keyManagers) {
```
Comment (sanjaypujare) [inline netty/src/main/java/io/grpc/netty/FixedTrustManagerFactory.java:30]: Name the parameter `trustManagers`

File: netty/src/main/java/io/grpc/netty/ProtocolNegotiators.java [hunk 8]
```diff
@@ -679,14 +731,6 @@ public TlsProtocolNegotiatorClientFactory(SslContext sslContext) {
...
     @Override public ProtocolNegotiator newNegotiator() {
-      SslContext sslContext = this.sslContext;
-      if (sslContext == null) {
-        try {
-          sslContext = GrpcSslContexts.forClient().build();
-        } catch (SSLException ex) {
-          throw new RuntimeException(ex);
-        }
-      }
       return tls(sslContext);
```
Comment (sanjaypujare) [inline netty/src/main/java/io/grpc/netty/ProtocolNegotiators.java:734]: Based on this change, should you have the `checkNotNull` check for `sslContext` on line 730?

File: okhttp/src/main/java/io/grpc/okhttp/OkHttpChannelBuilder.java [hunk 9]
```diff
@@ -537,14 +550,32 @@ static SslSocketFactoryResult sslSocketFactoryFrom(ChannelCredentials creds) {
         return SslSocketFactoryResult.error(
             "TLS features not understood: " + incomprehensible);
       }
-      SSLSocketFactory sslSocketFactory;
+      KeyManager[] km = null;
+      if (tlsCreds.getKeyManagers() != null) {
+        km = tlsCreds.getKeyManagers().toArray(new KeyManager[0]);
+      } else if (tlsCreds.getPrivateKey() != null) {
+        return SslSocketFactoryResult.error("byte[]-based private key unsupported. Use KeyManager");
```
Comment (sanjaypujare) [inline okhttp/src/main/java/io/grpc/okhttp/OkHttpChannelBuilder.java:557]: In a previous comment you said "Users tend to have bytes and not Certificate or PrivateKey. ...". We don't actually support that use case here. BTW `io.grpc.xds.internal.sds.trust.CertificateUtils.getPrivateKey` converts a byte array to a `PrivateKey` and it doesn't depend on netty...

File: okhttp/src/test/java/io/grpc/okhttp/OkHttpChannelBuilderTest.java [hunk 10]
```diff
@@ -155,6 +165,110 @@ public void sslSocketFactoryFrom_unsupportedTls() {
...
+          socket.getSession(); // Force handshake
+          serverSocket.set(socket);
+          serverListenSocket.close();
+        } catch (Throwable t) {
+          serverSocket.setException(t);
+        }
+      }
+    }).start();
+
+    ChannelCredentials creds = TlsChannelCredentials.newBuilder()
```
Comment (sanjaypujare) [inline okhttp/src/test/java/io/grpc/okhttp/OkHttpChannelBuilderTest.java:196]: How about a negative test also: client doesn't trust the server cert i.e. trustManager not initialized with `cert`

File: okhttp/src/test/java/io/grpc/okhttp/OkHttpChannelBuilderTest.java [hunk 11]
```diff
@@ -155,6 +165,110 @@ public void sslSocketFactoryFrom_unsupportedTls() {
...
```
Comment (sanjaypujare) [inline okhttp/src/test/java/io/grpc/okhttp/OkHttpChannelBuilderTest.java:245]: How about a couple of negative tests:

- `setNeedClientAuth(true)` on the server side but client doesn't send a cert
- client sends a cert that is not trusted by server

File: netty/src/test/java/io/grpc/netty/ProtocolNegotiatorsTest.java [hunk 12]
```diff
@@ -300,6 +324,128 @@ public ChannelCredentials withoutBearerTokens() {
...
+    InternalChannelz.Tls tls = expectSuccessfulHandshake(channelCreds, serverCreds);
+    assertThat(tls.remoteCert).isNull();
+  }
+
+  @Test
+  public void from_tls_mtls_require() throws Exception {
+    ServerCredentials serverCreds = TlsServerCredentials.newBuilder()
+        .keyManager(server1Cert, server1Key)
+        .trustManager(caCert)
+        .clientAuth(TlsServerCredentials.ClientAuth.REQUIRE)
```
Comment (sanjaypujare) [inline netty/src/test/java/io/grpc/netty/ProtocolNegotiatorsTest.java:376]: How about a negative test with `ClientAuth.REQUIRE` where the client doesn't present a certificate?

**Review by sanjaypujare — COMMENTED** [review]
Looks good. Will re-review after the current comments are addressed...

**Review by sanjaypujare — APPROVED** [review]
Looks great!
